src/components/svelte-web.ts

```diff
@@ -0,0 +1,226 @@
+    listeners = new Map<string, Map<Callback, Callback>>()
+    #component: SvelteComponent
+    get component() {
+      return this.#component
+    }
+
+    set component(value) {
+      // We need to make sure that when we recreate the component (as in the
+      // case of slots changing) that we copy over all of the event listeners.
+      this.#component = value
+      for (const [event, listeners] of this.listeners.entries()) {
+        for (const [callback, remove] of listeners.entries()) {
+          remove()
+          this.addEventListener(event, callback)
+        }
+      }
+    }
+
+    static get observedAttributes() {
+      return attributes
+    }
+
+    constructor() {
+      super()
+
+      // Mount shadow - this is where we're going to render our Component.
+      const shadow = this.attachShadow({ mode })
+
+      let lastSlots = new Set()
+      const updateSlots = () => {
+        const slotsNames = Array.from(this.children).map((c) =>
+          c.getAttribute('slot')
+        )
+        // Add default slot
+        if (this.childNodes.length) slotsNames.push(undefined)
+
+        // Slots didn't change, so nothing to do here.
+        // The component needs to get created, at least once
+        if (
+          this.component &&
+          // If the size is the same, and every one of our last slots
+          // is present, then nothing has changed, and we don't need
+          // to do anything here.
+          lastSlots.size === slotsNames.length &&
+          slotsNames.every((s) => lastSlots.has(s))
+        ) {
+          return
+        }
+
+        // Update the last slots we have, so if they change we know to update them.
+        lastSlots = new Set(slotsNames)
+
+        // Create a dictionary of the slotName: <slot name={slotName}/>
+        const slots = slotsNames.reduce(
+          (prev, next) => ({
+            ...prev,
+            [next ?? 'default']: [() => createSlot(next)]
+          }),
+          {}
+        )
+
+        // If we've already created the component, we might have some
+        // existing props. We need to create a snapshot of the component
+        // so we can recreate it as faithfully as possible.
+        // Note: We might be able to do some additional hackery here
+        // to copy over even more information from $$.ctx and exactly
+        // maintain the component state!
+        const existingProps = Object.keys(this.component?.$$.props ?? {})
+          .map((k) => [k, this[k]])
+          .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {})
+
+        // If the component already exists, destroy it. This is,
+        // unfortunately, necessary as there is no way to update slotted
+        // content in the output Svelte compiles to. This is a problem
+        // even when not doing crazy things:
+        // https://github.com/sveltejs/svelte/issues/5312
+        if (this.component) {
+          this.component.$destroy()
+        }
```

Comment:
I wonder what side effects this can produce. I guess there shouldn't really be any for Leo's simple components that shouldn't do much on 'mount', or maintain state. Although we're about to introduce AlertCentre which seems to! (not via slots though thankfully)

Reply:
Definitely if a component is doing something on mount we're going to get some interesting side effects. Unfortunately I don't really see a way around this for now.

We should be able to maintain state between mounts without too much difficulty but yeah, side effects are going to be a bit of a footgun.